contracts/src/registry/ERC1155SingletonUpgradable.sol

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {IERC165} from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+import {IERC1155} from "@openzeppelin/contracts/token/ERC1155/IERC1155.sol";
+import {ERC1155Upgradeable} from "@openzeppelin/contracts-upgradeable/token/ERC1155/ERC1155Upgradeable.sol";
+import {IERC1155MetadataURI} from "@openzeppelin/contracts/token/ERC1155/extensions/IERC1155MetadataURI.sol";
+import {ContextUpgradeable} from "@openzeppelin/contracts-upgradeable/utils/ContextUpgradeable.sol";
+import {Arrays} from "@openzeppelin/contracts/utils/Arrays.sol";
+
+/**
+ * @title IERC1155SingletonUpgradeable
+ * @dev Interface for the ERC1155SingletonUpgradeable contract
+ */
+interface IERC1155SingletonUpgradeable is IERC1155 {
+    function ownerOf(uint256 id) external view returns (address owner);
+}
+
+/**
+ * @title ERC1155SingletonUpgradeable
+ * @dev ERC1155 implementation that supports only a single token per ID.
+ * Stores owner instead of balance to allow fetching ownership information for a tokenId via `ownerOf`.
+ * This is an upgradeable version of the ERC1155Singleton contract.
+ */
+abstract contract ERC1155SingletonUpgradeable is 
```

Comment:
Can the shared code in this contract and `ERC1155Singleton` be refactored into an abstract base class inherited by both implementations?